### ISSUE TITLE
When I use DDSHelloWorldExampleTCP, I am having trouble accepting data while mapping using routes [12574]

### ISSUE THREAD

**Issue by XiaomuWang** [eProsima/Fast-DDS#2225]
I used the code in this link (https://github.com/eProsima/Fast-DDS/tree/master/examples/C%2B%2B/DDS/HelloWorldExampleTCP)to test and found that the data was not accepted.

I received data on the server and sent data on the edge. The router mapped IP and port, such as 172.31.1.115:20401(accessible to the server)->192.168.1.204(host):5100 (accessible to the edge). Then I ran the following commands respectively:

```
. / DDSHelloWorldExampleTCP subscriber -a 172.31.1.115  -p 20401

. / DDSHelloWorldExampleTCP publisher -a 192.168.1.204 -p 5100
```

I use Telnet to confirm that 172.31.1.15:20401 (server) and 192.168.1.204:5100 (edge) are both running services like:
![bdda55a9ff3366833346576f6b2983d](https://user-images.githubusercontent.com/46257486/134500310-fd0fff4b-e860-4f32-a26c-d18460295e4b.png)

![4c2b5b4a8e3020472e0c1ebe2380046](https://user-images.githubusercontent.com/46257486/134500328-b626b16c-a1e9-4d3f-a5b0-9c23cc7f6d8f.png)

However, I could not receive data at the receiving end, as shown in the following figure:
![3b187a090bbf90f6dfdf706323c638d](https://user-images.githubusercontent.com/46257486/134500345-b9e24d26-93c0-43e6-9855-826244b43759.png)

## Expected Behavior
I want my server to accept the data！I hope you can take time out of your busy schedule to help solve the problem！Thank you very much！

**Comment by XiaomuWang** [comment]
Supplement：
The uniqueness of this scenario is that we use routing to map the host and port (IP :192.168.1.204 and port: 5100) of the edge computing machine. We use routing to map it to (IP :172.31.1.115 and port: 20401), because the server side can access this mapping address and port (IP :172.31.1.115 and port: 20401), we send data from the edge computing computer and receive data from the server side, but my server side cannot receive data.

**Comment by XiaomuWang** [comment]
Because the data is forwarded by the router between my edge computing computer and the server, the data can be received theoretically because of the TCP used.

**Comment by XiaomuWang** [comment]
What should we do if the subscriber's computer cannot access the publisher's host IP?

**Comment by XiaomuWang** [comment]
I currently test that the IP of the publisher must be host IP. Can't it just be a public IP that both can access?

**Comment by MiguelCompany** [comment]
@jparisu Would you mind having a look at this one?

**Comment by jparisu** [comment]
Hi @XiaomuWang 

I think the problem here is the configuration of the parameters of the example.
In the first place, the parameters required by the example are not the one you are setting. The parameters for the Publisher are the **Publisher public IP (-a) and public Port (-p)** and the Subscriber requires the **Publisher public IP (-a) and public Port (-p)** in order to connect to it.
Thus, your example should work like this:
```
. / DDSHelloWorldExampleTCP subscriber -a 192.168.1.204 -p 5100
. / DDSHelloWorldExampleTCP publisher -a 192.168.1.204 -p 5100
```

When using TCP in Fast DDS you have to set a Participant that will work as a TCP server and another one as TCP client (this is not related with Discovery Server, neither with Publishers or Subscribers in those Participants). In this example, the Publisher side _(DomainParticipant + Publisher + DataWriter)_ is configured to be the TCP server, and so the Subscriber _(DomainParticipant + Subscriber + DataReader)_ must connect to it (using the Publisher IP and Port).
Thus, first of all you should use the side `172.31.1.15:20401 (server)` as Publisher in order to test the example (obviously you can have a Subscriber in the server side, but you would have to change a bit the example code).
However, if both have access to each other, this should not be a problem.

**Comment by XiaomuWang** [comment]
Thank you very much for your answer！I have two questions:

1. Must the public IP address be the publisher's host IP? I publish using the host IP of the server, because the subscribers of the client side can access this host IP and receive messages, but I publish on the client side (using the host IP of the server side) instead. But my subscriber on the server side cannot accept the message.

2. You said that I can add subscribers on the server side, how should I modify this? Can you provide further help?Because my server cannot access the host IP of the client！ @jparisu

**Comment by jparisu** [comment]
Hi @XiaomuWang 

This configuration is not trivial, so I would try to explain myself:

Using the `DDSHelloWorldTCP` you must take into account several thigs:
- There are two main nodes in a TCP connection: `TCP server` and `TCP client`
- In order to use TCP in a DDS communication, the nodes must be configured one as a client and one as a server. How to configure this is explained here: https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/tcp/tcp.html#wan-or-internet-communication-over-tcpv4
    - `TCP server` node must create a `TCPv4TransportDescriptor` and set its WAN IP and port.
    - `TCP client` node must create a `TCPv4TransportDescriptor` AND a locator that contains the **server WAN IP and port**.
- In `DDSHelloWorldTCP` there are only two kind of nodes:
    - `Publisher` (invoked by `publisher` argument) creates a `TCP server` AND a `DataWriter`.
    - `Subscriber` (invoked by `subscriber` argument) creates a `TCP client` AND a `DataReader`.
- **IMPORTANT**: there is no relation between a `DataWriter` or `DataReader` and a `TCP server` or `client`. The difference for `TCP` is in the `DomainParticipant` configuration, while the difference for publishing and subscribing is in the entities that are created under this `DomainParticipant`. 

So, the important point here is that **Fast DDS** could be configured to publish and subscribe in both sides `client` and `server` BUT the `DDHelloWorldTCP` is just an example, and because how it is configured, only the `Publisher` could be in the `server` side.

And now, I will answer your questions:

1a. The IP must be the public IP. If you are under a NAT, the IP must be the external IP of the router. If you are in the same LAN (under the same router), the IP must be the host IP. 
1b. The IP must always be the on of the `TCP server`. Using `DDSHelloWorldTCP` this would be the `Publisher` one.
2. In the `HelloWorldPublisher.cpp` the instruction `writer_ = publisher_->create_datawriter(topic_, wqos, &listener_);` is the one that creates the `DataWriter`. Using the instruction `reader_ = subscriber_->create_datareader(topic_, rqos, &listener_);` you can create `DataReaders` (you need to create beforehand a subscriber, please check `HelloWorldSubscriber.cpp`.

If you are still not able to configure the scenario that you are working with, we may arrange a meeting to explain it by video call, and I think it would be much easier.
Please, feel free to contact our operations department: support@eProsima.com

**Comment by XiaomuWang** [comment]
Thank you very much for your detailed suggestions. I realized subscription on the server and publishing on the client! Finally meet the requirements of the scene, thank you again for your detailed suggestions! I'll close it for now!

**Comment by XiaomuWang** [comment]
When I use this use case for transmission, I find that the received data is much smaller than the sent data, as shown in the figure:
![f1](https://user-images.githubusercontent.com/46257486/135390252-23c22ec2-eeab-43c7-bbe2-db87e0a05f63.png)

But I changed the size of the message, as shown:
![f2](https://user-images.githubusercontent.com/46257486/135390292-acbd9e18-597b-4213-8c32-67acbafb8306.png)

@jparisu

**Comment by jparisu** [comment]
Hi @XiaomuWang. Happy to know that you are still on track.

I would need more information regarding your issue. 
First of all, I assume that you used **Fast DDS Gen** to generate the new `HelloWorld.cxx` `HelloWorld.h` `HelloWorldPubSubTypes.cxx` and `HelloWorldPubSubTypes.h`.
Secondly, and most important: how are you getting the "length" that you are printing? If you are using the `std::string::length` function, be aware that the length would be the size of the message, and not the actual size of the data sent through the network. Thus, if your message is only 10 chars long, that is the expected size of it.

**Comment by XiaomuWang** [comment]
Hi @jparisu Thank you very much for your answer!

First, I used Fast DDS Gen to generate the new code.
Secondly, in my code of **HelloWorldPublisher.cpp** and **HelloWorldSubscriber.cpp** , I use hello_.message().size() to print the length. Shouldn't this be the actual length of the sent data? I don't know what's going on.

Here is the link to my [code:](https://github.com/XiaomuWang/dds_tcp)
Compared to the sample code, I made the following changes:
1, the server side is changed to subscribe, the client side is changed to publish;
2. Modify helloworld. idl to increase the length of message, where index is the timestamp in the code;
3. Use Fast DDS Gen to generate new code。

**Comment by jparisu** [comment]
Ok, I think I know what is happening here.

First of all, using `hello_.message().size()` will not give you the exact size of the message you are receiving, but the size of the `std::string` that is being deserialized from it. Thus, if you create a message of size 2048, the size of the DATA sent through the network will be 2048, BUT if the string actually contains 10 char followed by `\0` , the deserialization of this data will create an `std::string` of size 10 (11 with `\0`, but it is not counted in `std::string::size()`).
Here you can see how the deserialization is being done: https://github.com/XiaomuWang/dds_tcp/blob/master/src/HelloWorld.cxx#L127
And here, what happens when you call `hello_.message().size(): `https://github.com/XiaomuWang/dds_tcp/blob/master/src/HelloWorld.cxx#L187

I would say that you are trying to send binary data inside a string, and thus, once it is received in the Subscriber side, every `\0` in the data will stop the deserialization.
In order to send binary data use array of bytes: `octet message[2048]` (be careful with the sides you are working with, if you set in your .idl a maximum size of 2048 and your actual string is bigger, the string will be cut when sent).

Finally, in order to retrieve the actual size that your serialized has (the actual size of data that is sent through the cable) use the method `HelloWorld::getCdrSerializedSize(hello_)`, and this should be the same in both sides (unless the case with strings and `\0` as we have mentioned).

**Comment by MiguelCompany** [comment]
> use array of bytes: `octet message[2048]`

That one will always transmit 2048 bytes. Depending on the use case, it may even be better to use `sequence<octet, 2048>`, in which case the date will be a sequence of up to 2048 bytes.

**Comment by XiaomuWang** [comment]
Hi!Thank you very much! @jparisu @MiguelCompany I understand what you mean, thank you very much for your patience. I finally understand why only part of the string can be sent. In addition, when I use the data structure you mentioned (**sequence< octet, 1024**>), I encounter a problem that is difficult to solve, and the Subscriber will print Subscriber matched. But **I did not print the received data**, and **the CDR size I received is smaller** than the one I received, what is the reason?
This is the message printed by my subscribers:
![sub](https://user-images.githubusercontent.com/46257486/136682520-798f5997-802d-4f29-b67d-a9b02a5376fc.png)
This is the message printed by my publisher:
![pub](https://user-images.githubusercontent.com/46257486/136682534-46dfdbac-07ad-49c7-99f7-d29c7f3e3a1f.png)
In addition, I would like to explain that I use the server as the subscriber, use the client as the publisher, and I made some changes to the main function, so I need to change the IP and port number in the main function!
Here's the link to my code:https://github.com/XiaomuWang/dds_seq_test
This bug is really difficult for me to solve! I hope you can help me solve this bug in your busy schedule. If you can solve it, I will be very grateful!

**Comment by XiaomuWang** [comment]
It's worth mentioning that the code was just modified to use a string structure to successfully receive data, but the data couldn't be deserialized when it hit \0, and I still can't send my data in its entirety.

**Comment by XiaomuWang** [comment]
@jparisu Hello, I use HelloWorldExampleTCP code to transfer sequence types, and I find that as the data gets bigger and bigger, the transfer gets slower and slower until it finally stops, where should I optimize？
![image](https://user-images.githubusercontent.com/46257486/137840301-480ee9a0-ab6d-4e59-b566-eb49eeb282cb.png)
I've set it up：wparam.historyMemoryPolicy = rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
Since my biggest data was around 700K and the transfer was slow, I set it to 800000, but I found that the transfer was slow at around 500K！

**Comment by XiaomuWang** [comment]
Occasionally this problem occurs, and then subscribers can continue to receive it.

**Comment by XiaomuWang** [comment]
For your convenience, I have tested on our TCP demo, and the result is the same. After printing Subscriber matched, issue and acceptance will be stopped, but after acceptance is stopped, the publisher will continue to issue data. Is it because the amount of data is too large and exceeds the performance of FAST DDS?![699d4c5f3770abbbfd2213639135323](https://user-images.githubusercontent.com/46257486/137862130-91ac6dee-d53d-4cb6-9ce9-c9750d7c39f3.png)!
Publisher side:
![447949027aff7426ba28a391f36c09e](https://user-images.githubusercontent.com/46257486/137862110-ca64fbd5-5165-434b-9b85-e8d3c132e661.png)
Subscriber side:
![699d4c5f3770abbbfd2213639135323](https://user-images.githubusercontent.com/46257486/137862219-d77dd2aa-1e05-4a23-a136-15110c80d79a.png)

fast dds version：2.3.0

**Comment by jparisu** [comment]
Hi @XiaomuWang 

Glad to know that you are still working with Fast DDS.

First of all, it seems like you have changed the API that your are using.
You have to be aware that Fast DDS supports 2 different APIs. FastPubSub (deprecated) was the old one. Fast DDS is the new and the maintained one, and it is available since Fast DDS v2.0
In the examples directory, there is a subdirectory called `DDS` with examples using the new API (https://github.com/eProsima/Fast-DDS/tree/master/examples/C%2B%2B/DDS) while in the upper directory there are old examples with the old API (https://github.com/eProsima/Fast-DDS/tree/master/examples/C%2B%2B).
I realized you are using the old API, and this was unexpected as at the beginning of this thread you were using the new DDS API.

Another advice is to check the communication flows that are available with the new API that avoid copies when communicating two processes in the same host (don't know if this is your case): 
- The easiest one and used by default is Shared Memory: https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/shared_memory/shared_memory.html
- Also, with an small refactor of your data type (your .idl file) you can have a bounded data type, and you would be able to use Data Sharing by default: https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/datasharing.html
- And with a small refactor of the API in writers and readers you could get the most efficient communication flow Zero Copy https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/zero_copy/zero_copy.html#use-case-zero-copy. For this purpose check example: https://github.com/eProsima/Fast-DDS/tree/master/examples/C%2B%2B/DDS/ZeroCopyExample

In order to further help you with your use case, please check your internet connection speed, and let us know which is your experiment architecture (if every process is running in the same host, or you are communicating different hosts).
Please, facilitate me the arguments that you use to execute the Publisher (are not seen in the screenshot).
Be aware that the slow down that you are experienced could be because of the saturation of your network interface rather than because of Fast DDS.

BTW: It seems like you overcame the problem about the CDR size message, right?

**Comment by XiaomuWang** [comment]
Hi @jparisu .I am very glad to see your timely reply, because the effect of transmitting this data on the new API was not good before, and I did a simple test on the new API today(code [link).](https://github.com/XiaomuWang/NewAPIHelloWorldExampleTCP)

To answer your question first:
1. I communicate on two hosts, and the communication network between hosts is stable;
2. I have tested both the new API and the old API, but the normal transmission of my data cannot be realized.
3. In addition, I am very happy to tell you that I have solved the problem of CDR message size and I can print it out.

Here are some details about my implementation on the new API:
Execute command:
The result of the execution is an error and the data is stopped write after the match is successful (note that the length of the vector is printed) :
1. Publisher side :
`. / DDSHelloWorldExampleTCP 10.10.48.101 publisher - a - p 5555`
2. The publish side print information:
root@adminroot:/workspace/init_test/baoan_test/new_API/HelloWorldExampleTCP/build# ./DDSHelloWorldExampleTCP publisher -a 10.10.48.101 -p 5555
Starting
10.10.48.101:5555
2021-10-19 16:50:13.045 [DATASHARING_PAYLOADPOOL Error] Failed to create segment fast_datasharing_01.0f.9b.44.62.04.a6.19.01.00.00.00_0.0.1.3: No such file or directory -> Function init_shared_memory
2021-10-19 16:50:13.045 [RTPS_WRITER Error] Could not initialize DataSharing writer pool -> Function init
Publisher running. Please press enter to stop_ the Publisher at any time.
[RTCP] Message: 780380 with index: 1 SENT
[RTCP] Message: 780380 with index: 2 SENT
[RTCP] Message: 780380 with index: 3 SENT
[RTCP] Message: 780380 with index: 4 SENT
[RTCP] Message: 780380 with index: 5 SENT
[RTCP] Message: 780380 with index: 6 SENT
[RTCP] Message: 780380 with index: 7 SENT
[RTCP] Message: 780380 with index: 8 SENT
[RTCP] Message: 780380 with index: 9 SENT
[RTCP] Message: 780380 with index: 10 SENT
[RTCP] Message: 780380 with index: 11 SENT
[RTCP] Message: 780380 with index: 12 SENT
[RTCP] Message: 780380 with index: 13 SENT
[RTCP] Message: 780380 with index: 14 SENT
[RTCP] Message: 780380 with index: 15 SENT
[RTCP] Message: 780380 with index: 16 SENT
[RTCP] Message: 780380 with index: 17 SENT
[RTCP] Message: 780380 with index: 18 SENT
[RTCP] Message: 780380 with index: 19 SENT
[RTCP] Message: 780380 with index: 20 SENT
[RTCP] Message: 780380 with index: 21 SENT
[RTCP] Message: 780380 with index: 22 SENT
[RTCP] Message: 780380 with index: 23 SENT
[RTCP] Message: 780380 with index: 24 SENT
[RTCP] Message: 780380 with index: 25 SENT
[RTCP] Message: 780380 with index: 26 SENT
[RTCP] Message: 780380 with index: 27 SENT
[RTCP] Message: 780380 with index: 28 SENT
[RTCP] Message: 780380 with index: 29 SENT
[RTCP] Message: 780380 with index: 30 SENT
[RTCP] Message: 780380 with index: 31 SENT
[RTCP] Message: 780380 with index: 32 SENT
[RTCP] Message: 780380 with index: 33 SENT
[RTCP] Message: 780380 with index: 34 SENT
[RTCP] Message: 780380 with index: 35 SENT
[RTCP] Message: 780380 with index: 36 SENT
[RTCP] Message: 780380 with index: 37 SENT
[RTCP] Message: 780380 with index: 38 SENT
[RTCP] Message: 780380 with index: 39 SENT
[RTCP] Message: 780380 with index: 40 SENT
[RTCP] Message: 780380 with index: 41 SENT
[RTCP] Message: 780380 with index: 42 SENT
[RTCP] Message: 780380 with index: 43 SENT
[RTCP] Message: 780380 with index: 44 SENT
[RTCP] Message: 780380 with index: 45 SENT
[RTCP] Message: 780380 with index: 46 SENT
[RTCP] Message: 780380 with index: 47 SENT
[RTCP] Message: 780380 with index: 48 SENT
[RTCP] Message: 780380 with index: 49 SENT
[RTCP] Message: 780380 with index: 50 SENT
[RTCP] Message: 780380 with index: 51 SENT
[RTCP] Message: 780380 with index: 52 SENT
[RTCP] Message: 780380 with index: 53 SENT
[RTCP] Message: 780380 with index: 54 SENT
[RTCP] Message: 780380 with index: 55 SENT
[RTCP] Publisher matched
[RTCP] Message: 780380 with index: 56 SENT

3. Subscribe side(on other computer):
. / DDSHelloWorldExampleTCP 10.10.48.101 subscriber - a - p 5555
4.The  subscribe side print information:
root@IDC_GPU_Server-1:/workspace/temp/HelloWorldExampleTCP/build# ./DDSHelloWorldExampleTCP subscriber -a 10.10.48.101 -p 5555
Starting
10.10.48.101:5555
[RTCP] Subscriber running. Please press enter to stop the Subscriber
[RTCP] Subscriber matched
Message 780380 27 RECEIVED
Message 780380 28 RECEIVED
Message 780380 29 RECEIVED
Message 780380 30 RECEIVED
Message 780380 31 RECEIVED
Message 780380 32 RECEIVED
Message 780380 33 RECEIVED
Message 780380 34 RECEIVED
Message 780380 35 RECEIVED


##Current Behavior##
1. The above error message is displayed.
2. The old API and the new API have a common problem when sending large vectors: once a match is made, the publisher stops writing data.

I would like to explain my problem clearly first, and I will try some of the suggestions you mentioned above. Of course, you can put forward any new ideas at any time!
